File: packages/core/src/ai/ai-service.ts
import type { AIProviderPublicConfig, CommentInput, CommentRecord, IntentResult, KeywordPlan, LeadRecord, LeadScore, SearchResult } from '../domain/types.ts'
import type { LLMClient } from './llm-client.ts'

const intentKeywords = {
  high: ['想买', '多少钱', '哪里买', '求链接', '怎么买', 'price', 'buy', 'link'],
  medium: ['好用吗', '值得买吗', '推荐吗', '对比', '想了解', '了解一下', '有优惠', 'review', 'worth', 'best'],
  low: ['收藏', '看看', '喜欢', '种草', '不错', 'mark', 'interesting']
}

export class AIService {
  private provider?: AIProviderPublicConfig
  private apiKey?: string
  private llmClient?: LLMClient

  constructor(llmClient?: LLMClient) {
    this.llmClient = llmClient
  }

  configureProvider(config?: AIProviderPublicConfig, apiKey?: string): void {
    this.provider = config
    this.apiKey = apiKey
  }

  setLLMClient(client?: LLMClient): void {
    this.llmClient = client
  }

  currentProvider(): AIProviderPublicConfig | undefined {
    return this.provider
  }

  expandKeywords(seed: string, locales = ['zh-CN']): KeywordPlan {
    const base = seed.trim()
    if (!base) return { seed, keywords: [], locales }
    const keywords = [
      base,
      `${base} 推荐`,
      `${base} 怎么选`,
      `${base} 避坑`,
      `${base} 价格`,
      `${base} 测评`,
      `${base} 求链接`
    ]
    if (locales.some((locale) => locale.toLowerCase().startsWith('en'))) {
      keywords.push(`${base} review`, `best ${base}`, `${base} price`, `where to buy ${base}`)
    }
    return { seed: base, keywords: [...new Set(keywords)], locales }
  }

  rankSearchResults(keyword: string, results: SearchResult[]): SearchResult[] {
    const key = keyword.trim().toLowerCase()
    return results
      .map((result) => {
        const haystack = `${result.title} ${result.snippet}`.toLowerCase()
        let relevance = result.relevance
        if (key && haystack.includes(key)) relevance += 0.5
        if (/(推荐|测评|怎么买|价格|review|best|buy)/i.test(haystack)) relevance += 0.2
        return { ...result, relevance: Math.min(1, relevance) }
      })
      .sort((a, b) => b.relevance - a.relevance)
  }

  analyzeIntent(comment: CommentInput): IntentResult {
    const text = comment.text.toLowerCase()
    for (const level of ['high', 'medium', 'low'] as const) {
      const matched = intentKeywords[level].filter((keyword) => text.includes(keyword.toLowerCase()))
      if (matched.length > 0) {
        return {
          level,
          confidence: level === 'high' ? 0.86 : level === 'medium' ? 0.72 : 0.55,
          keywords: matched,
          reason: `命中${level}意向关键词`
        }
      }
    }
    return { level: 'none', confidence: 0.3, keywords: [], reason: '未发现明确购买信号' }
  }

  async analyzeIntentWithModel(comment: CommentInput): Promise<IntentResult> {
    return (await this.analyzeIntentWithSource(comment)).intent
  }

  async analyzeIntentWithSource(comment: CommentInput): Promise<{ intent: IntentResult; source: 'model' | 'rule' }> {
    if (!this.provider?.enabled || this.provider.provider === 'rule' || !this.apiKey || !this.llmClient) {
      return { intent: this.analyzeIntent(comment), source: 'rule' }
    }
    try {
      return { intent: await this.llmClient.analyzeIntent(comment, { provider: this.provider, apiKey: this.apiKey }), source: 'model' }
    } catch {
      return { intent: this.analyzeIntent(comment), source: 'rule' }
    }
  }

  scoreLead(intent: IntentResult, likes = 0): LeadScore {
    const base = { high: 90, medium: 65, low: 35, none: 0 }[intent.level]
    const likeBoost = likes >= 10 ? 5 : 0
    const keywordBoost = intent.keywords.length > 0 ? 5 : 0
    const score = Math.min(100, base + likeBoost + keywordBoost)
    return {
      score,
      level: intent.level,
      reason: [
        intent.reason,
        `基础分 ${base}`,
        keywordBoost ? `关键词加分 ${keywordBoost}` : '',
        likeBoost ? `互动加分 ${likeBoost}` : ''
      ].filter(Boolean).join('；'),
      suggestedAction: score >= 80 ? '优先跟进' : score >= 50 ? '加入跟进池' : '低优先级观察'
    }
  }

  commentToLead(comment: CommentRecord): LeadRecord | null {
    const intent = this.analyzeIntent({
      platformKey: comment.platformKey,
      contentUrl: comment.contentUrl,
      nickname: comment.nickname,
      text: comment.text,
      likes: comment.likes
    })
    if (intent.level === 'none') return null
    const score = this.scoreLead(intent, comment.likes)
    return {
      id: `lead-${comment.id}`,
      commentId: comment.id,
      platformKey: comment.platformKey,
      contentId: comment.contentId,
      nickname: comment.nickname,
      text: comment.text,
      intentLevel: intent.level,
      confidence: intent.confidence,
      keywords: intent.keywords,
      score: score.score,
      scoreReason: score.reason,
      suggestedAction: score.suggestedAction,
      status: 'new',
      createdAt: new Date().toISOString()
    }
  }

  async commentToLeadAsync(comment: CommentRecord): Promise<LeadRecord | null> {
    return (await this.commentToLeadWithMeta(comment)).lead
  }

  async commentToLeadWithMeta(comment: CommentRecord): Promise<{ lead: LeadRecord | null; source: 'model' | 'rule' }> {
    const result = await this.analyzeIntentWithSource({
      platformKey: comment.platformKey,
      contentUrl: comment.contentUrl,
      nickname: comment.nickname,
      text: comment.text,
      likes: comment.likes
    })
    const intent = result.intent
    const source = result.source
    if (intent.level === 'none') return { lead: null, source }
    const score = this.scoreLead(intent, comment.likes)
    return { lead: {
      id: `lead-${comment.id}`,
      commentId: comment.id,
      platformKey: comment.platformKey,
      contentId: comment.contentId,
      nickname: comment.nickname,
      text: comment.text,
      intentLevel: intent.level,
      confidence: intent.confidence,
      keywords: intent.keywords,
      score: score.score,
      scoreReason: score.reason,
      suggestedAction: score.suggestedAction,
      status: 'new',
      createdAt: new Date().toISOString()
    }, source }
  }
}

File: core/ai_service.py
"""
统一 AI 服务层。

现有 LLM 类主要服务评论意向识别。该服务层把未来的关键词扩展、搜索结果
重排、线索评分、话术生成等能力集中到一个稳定入口，避免 UI 直接依赖各厂商模型。
"""
from __future__ import annotations

from dataclasses import dataclass
from typing import Iterable, Optional

from llm.base import BaseLLM
from storage.models import LeadInfo, SearchResult


@dataclass
class LeadScore:
    score: int
    reason: str = ""
    suggested_action: str = ""


class AIService:
    """业务级 AI 能力门面。"""

    def __init__(self, llm: Optional[BaseLLM] = None):
        self.llm = llm

    def expand_keywords(self, keyword: str, locale: str = "zh-CN") -> list[str]:
        """生成搜索词扩展；无模型时使用可解释的规则兜底。"""
        base = keyword.strip()
        if not base:
            return []
        expansions = [
            base,
            f"{base} 推荐",
            f"{base} 怎么选",
            f"{base} 避坑",
            f"{base} 价格",
            f"{base} 好用吗",
            f"{base} 测评",
            f"{base} 求链接",
        ]
        if locale.lower().startswith("en"):
            expansions.extend([f"{base} review", f"best {base}", f"{base} price"])
        return list(dict.fromkeys(expansions))

    def rank_search_results(self, keyword: str, results: Iterable[SearchResult]) -> list[SearchResult]:
        """搜索结果排序。当前先用规则评分，后续可接 LLM/embedding。"""
        key = keyword.lower().strip()
        ranked: list[SearchResult] = []
        for result in results:
            text = f"{result.title} {result.snippet}".lower()
            score = result.relevance or 0.0
            if key and key in text:
                score += 0.5
            if any(signal in text for signal in ("推荐", "测评", "怎么买", "价格", "review", "best")):
                score += 0.2
            result.relevance = score
            ranked.append(result)
        return sorted(ranked, key=lambda item: item.relevance, reverse=True)

    def score_lead(self, lead: LeadInfo) -> LeadScore:
        """线索评分，先基于现有字段给出稳定兜底。"""
        intent_value = lead.intent_level.value if hasattr(lead.intent_level, "value") else str(lead.intent_level)
        base_scores = {"高": 90, "中": 65, "低": 35, "无": 0}
        score = base_scores.get(intent_value, 0)
        if lead.likes > 10:
            score += 5
        if lead.intent_keywords:
            score += 5
        score = max(0, min(100, score))
        action = "优先跟进" if score >= 80 else "加入跟进池" if score >= 50 else "低优先级观察"
        return LeadScore(score=score, reason=f"意向等级为{intent_value}", suggested_action=action)

    def generate_followup(self, lead: LeadInfo, product_info: str = "") -> str:
        """生成跟进话术。模型不可用时返回模板话术。"""
        text = lead.comment_text or ""
        intent_value = lead.intent_level.value if hasattr(lead.intent_level, "value") else str(lead.intent_level)
        if self.llm:
            try:
                return self.llm.generate_follow_up(text, intent_value, product_info)
            except Exception:
                pass
        nickname = lead.nickname or "您好"
        return f"{nickname}，看到您关注这个产品，我可以把价格、型号和购买方式整理给您，方便您对比选择。"

    def summarize_batch(self, leads: Iterable[LeadInfo]) -> str:
        leads = list(leads)
        total = len(leads)
        high = sum(1 for lead in leads if getattr(lead.intent_level, "value", lead.intent_level) == "高")
        medium = sum(1 for lead in leads if getattr(lead.intent_level, "value", lead.intent_level) == "中")
        return f"本批次共 {total} 条线索，其中高意向 {high} 条，中意向 {medium} 条。"
